### PR TITLE
Modtop

### PR DESCRIPTION
**Pull request by thepineapplepirate**

### CODE REVIEW

File: tools/buildtools/topologyeditors/add_top.xml [hunk 1]
```diff
@@ -0,0 +1,129 @@
+<tool id="modify_topology" name="Adding New Topology Information" version="@TOOL_VERSION@+galaxy@GALAXY_VERSION@">
```
Comment (bgruening) [inline tools/buildtools/topologyeditors/add_top.xml:1]: add profile="20.09"

File: tools/buildtools/topologyeditors/add_top.xml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    <description>to a GROMACS topology file</description>
+
+    <macros>
+        <import>macros.xml</import>
+        <token name="@GALAXY_VERSION@">0</token>
+    </macros>
+
+    <expand macro="requirements">
+       <requirement type="package" version="3.8">python</requirement>
```
Comment (bgruening) [inline tools/buildtools/topologyeditors/add_top.xml:10]: why not 3.9?

File: tools/buildtools/topologyeditors/add_top.xml [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    </expand>
+
+    <command detect_errors="exit_code"><![CDATA[
+        #if $str($functionality.what2add) == "mol":
+            python3 $__tool_directory__/newmolparameters.py --top_file '$functionality.inputtopology' --atom_file '$functionality.nonbondparam' --mol_file '$functionality.bondparam' --out '$output1'
```
Comment (bgruening) [inline tools/buildtools/topologyeditors/add_top.xml:15]: please single-quote all paths ... e.g. `'$__tool_directory__/newmolparameters.py'`

and below

File: tools/buildtools/topologyeditors/add_top.xml [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                 <param name="nonbondparam" type="data" format="txt" label="Atom Types with Nonbonded Parameters" help="Atom types information from grep"/>
+                 <param name="bondparam" type="data" format="txt,itp" label="Molecule type information with bonded parameters" help="molecule type information"/>
+                 <param name="posres" type="data" format="itp" label="position restraint file" help="Position restraint file created previously"/>
+                 <param name="targetmolecule" type="text" label="Target molecule type" help="The molecule type name to which the restraints were applied"/>
+             </when>
+
+        </conditional>  
+    </inputs>
+    <outputs>
+        <data name="output1" format="top" label="Modified file with new topology information ${on_string}">
```
Comment (bgruening) [inline tools/buildtools/topologyeditors/add_top.xml:58]: Can you give it a more meaningful name than output1 ... those names are used later in workflows etc ...

File: tools/buildtools/topologyeditors/add_top.xml [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        </data>
+    </outputs>
+    <tests>
+        <test>
+            <param name="functionality|what2add" value="mol" />
```
Comment (bgruening) [inline tools/buildtools/topologyeditors/add_top.xml:63]: can you use the more explicit synatx here:

<conditional name="">
    <param name="what2add" value="mol" />
    <param name="inputtopology" value="cid1_GMX.top" />
    ....
</conditional>

File: tools/buildtools/topologyeditors/extract_top.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                        '--out_bondparam', default=None,
+                        help='moleculetype section')
+    parser.add_argument(
+                        '--out_nonbondparam', default=None,
+                        help='atomtypes section')
+
+    args = parser.parse_args()
+    # extracts the atom types with nonbonded terms from
+    # the new molecules and puts them in a new file
+    inFile = open(args.top_file)
```
Comment (bgruening) [inline tools/buildtools/topologyeditors/extract_top.py:21]: can you use `with open() as inFile` here?

File: tools/buildtools/topologyeditors/extract_top.xml [hunk 7]
```diff
@@ -0,0 +1,69 @@
+<tool id="extract_topology" name="Extracting Topology Information" version="@TOOL_VERSION@+galaxy@GALAXY_VERSION@">
```
Comment (bgruening) [inline tools/buildtools/topologyeditors/extract_top.xml:1]: profile

File: tools/buildtools/topologyeditors/extract_top.xml [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    <description>from a GROMACS topology file</description>
+
+    <macros>
+        <import>macros.xml</import>
+        <token name="@GALAXY_VERSION@">0</token>
+    </macros>
+
+    <expand macro="requirements">
+       <requirement type="package" version="3.8">python</requirement>
```
Comment (bgruening) [inline tools/buildtools/topologyeditors/extract_top.xml:10]: you have python already in the macro

File: tools/buildtools/topologyeditors/macros.xml [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+    <token name="@TOOL_VERSION@">0</token>
+    <xml name="requirements">
+        <requirements>
+            <requirement type="package" version="3.8">python</requirement>
+            <yield />
+        </requirements>
+    </xml>
+    <xml name="citations">
+        <citations>
+            <citation type="doi">10.1016/j.softx.2015.06.001</citation>
```
Comment (bgruening) [inline tools/buildtools/topologyeditors/macros.xml:11]: here is no yield, so you can not expand it and add new citations in your tool, but you are trying to do it

**Review by bgruening — COMMENTED** [review]
Thanks @thepineapplepirate.

Small review inline.

File: tools/buildtools/topologyeditors/add_top.xml (hunk 2 above)
Comment (thepineapplepirate) [reply tools/buildtools/topologyeditors/add_top.xml:10]: no particular reason, but I can change it to 3.9

File: tools/buildtools/topologyeditors/add_top.xml (hunk 5 above)
Comment (thepineapplepirate) [reply tools/buildtools/topologyeditors/add_top.xml:63]: not sure I understand what you are referring to

File: tools/buildtools/topologyeditors/add_top.xml (hunk 2 above)
Comment (thepineapplepirate) [reply tools/buildtools/topologyeditors/add_top.xml:10]: I think I was still using 3.8, and just added that as a minimum req.

File: tools/buildtools/topologyeditors/add_top.xml (hunk 1 above)
Comment (thepineapplepirate) [reply tools/buildtools/topologyeditors/add_top.xml:1]: added

File: tools/buildtools/topologyeditors/add_top.xml (hunk 2 above)
Comment (thepineapplepirate) [reply tools/buildtools/topologyeditors/add_top.xml:10]: changed to 3.9, and left this only in the macros.xml file, and removed this line from add_top.xml and extract_top.xml.

File: tools/buildtools/topologyeditors/add_top.xml (hunk 5 above)
Comment (thepineapplepirate) [reply tools/buildtools/topologyeditors/add_top.xml:63]: ?

File: tools/buildtools/topologyeditors/add_top.xml (hunk 4 above)
Comment (thepineapplepirate) [reply tools/buildtools/topologyeditors/add_top.xml]: changed new name to 'newtop'

File: tools/buildtools/topologyeditors/extract_top.py (hunk 6 above)
Comment (thepineapplepirate) [reply tools/buildtools/topologyeditors/extract_top.py]: changed to requested format

File: tools/buildtools/topologyeditors/extract_top.xml (hunk 7 above)
Comment (thepineapplepirate) [reply tools/buildtools/topologyeditors/extract_top.xml]: made change

File: tools/buildtools/topologyeditors/extract_top.xml (hunk 8 above)
Comment (thepineapplepirate) [reply tools/buildtools/topologyeditors/extract_top.xml]: removed

File: tools/buildtools/topologyeditors/add_top.xml (hunk 3 above)
Comment (thepineapplepirate) [reply tools/buildtools/topologyeditors/add_top.xml]: made change

File: tools/buildtools/topologyeditors/add_top.xml (hunk 5 above)
Comment (thepineapplepirate) [reply tools/buildtools/topologyeditors/add_top.xml]: fixed

**Comment by bgruening** [comment]
@thepineapplepirate this directory does not contain a .shed.yml file, left?

**Comment by thepineapplepirate** [comment]
> @thepineapplepirate this directory does not contain a .shed.yml file, left?

I do not believe I made one. Let me know if it is needed.

**Comment by bgruening** [comment]
Yes, we need one, otherwise test will not run and the tools will not be uploaded to the toolshed :)

**Comment by thepineapplepirate** [comment]
> Yes, we need one, otherwise test will not run and the tools will not be uploaded to the toolshed :)

Ok, will write one up :) 👍

**Review by bgruening — COMMENTED** [review]
Should we prefix the tool IDs with gromacs?

**Comment by thepineapplepirate** [comment]
> Should we prefix the tool IDs with gromacs?

i think you're referring to the python scripts and their names starting with 'gmxtras'. the reason we did this is because these are going to be part of a new series of scripts we're writing to help with a number of extra things that could be useful as auxiliary tools in GROMACS, and we're setting up a repo just for the python scripts called 'gmxtras' (see link below), so for consistency we figured it would be best to keep the names the same. let me know if that's cool :).

https://github.com/BlankenbergLab/gmxtras

**Comment by bgruening** [comment]
I'm talking about: `<tool id="modify_topology"`.... modify_topology is very generic imho.

We are also administrating a Galaxy and those IDs are everywhere, so the more descriptive they are the better.

**Comment by thepineapplepirate** [comment]
> I'm talking about: `<tool id="modify_topology"`.... modify_topology is very generic imho.
> 
> We are also administrating a Galaxy and those IDs are everywhere, so the more descriptive they are the better.

Ok, I see. Sorry i was confused. Are you cool with "gromacs_modify_topology" and "gromacs_extract_topology" (or gmx, instead of gromacs). let me know, will make the change and push it back.

**Comment by bgruening** [comment]
gromacs is ok for me :)

**Comment by thepineapplepirate** [comment]
> gromacs is ok for me :)

sweet! changes made and pushed back.

**Review by bgruening — APPROVED** [review]

**Pull request merged**